subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r51/ProjectConfigurationProgressEventCrossVersionSpec.groovy

```diff
@@ -313,6 +313,42 @@ class ProjectConfigurationProgressEventCrossVersionSpec extends ToolingApiSpecif
         result.duration >= Duration.ofMillis(sleepMillis)
     }
 
+    def "only counts execution time of container callbacks once"() {
+        given:
+        def sleepDurationMillis = 500
+        file("build.gradle") << """
+            configurations {
+                foo
+            }
+
+            apply plugin: MyPlugin
+
+            class MyPlugin implements Plugin<Project> {
+                void apply(Project project) {
+                    project.configurations.all {
+                        if (name == 'foo') {
+                            long start = System.currentTimeMillis()
+                            while (System.currentTimeMillis() - start < $sleepDurationMillis) {
+                                Thread.sleep(${sleepDurationMillis / 2})
+                            }
+                        }
+                    }
+                }
+            }
+        """
+
+        when:
+        runBuild("tasks", EnumSet.of(OperationType.PROJECT_CONFIGURATION)) {
+            it.withArguments("-D${CollectionCallbackActionDecorator.CALLBACK_EXECUTION_BUILD_OPS_TOGGLE}=true")
+        }
+
+        then:
+        def pluginResults = getPluginConfigurationOperationResult(":").getPluginConfigurationResults()
+        def result = pluginResults.find { it.plugin.displayName.contains("MyPlugin") }
+        result.duration >= Duration.ofMillis(sleepDurationMillis)
+        result.duration < Duration.ofMillis(2 * sleepDurationMillis)
```

Comment:
This may become flaky if the build server is slow/overloaded. I wonder if we should stick with a unit test for this one.

Reply:
I'll keep an eye on it.